### PR TITLE
silo: fix include bug in external mpi

### PR DESCRIPTION
**Pull request by junghans**
Found here: https://travis-ci.org/junghans/proxy-ci/jobs/342463086

cc @markcmiller86

### CODE REVIEW

**Comment by junghans** [comment]
The reason is that Ubuntu has its `mpi.h` in `'/usr/lib/x86_64-linux-gnu/openmpi/include' (not `/usr/include`).

**Comment by junghans** [comment]
@tgamblin @adamjstewart so I guess the real question is, is there a way to specify the `include` path of an external package independently of the `prefix` in `paths:` block of `packages.yaml`?

**Comment by markcmiller86** [comment]
@junghans ... I am having trouble grocking the original error log demonstrating the problem. Can you summarize? Is this Silo or MACSio's link to Silo

**Comment by markcmiller86** [comment]
Ok, so I looked a bit further and am thinking this is the smoking gun...

```
| #include <hdf5.h>
configure:28803: result: no
configure:28836: checking for hdf5.h
configure:28845: result: no
configure:28869: checking for H5open
configure:28933: /home/travis/build/junghans/proxy-ci/spack/lib/spack/env/gcc/gcc -o conftest -fPIC -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wdeclaration-after-statement -I/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-crbbyudcj22o5onhxloe32v5d2nkwa5w/include  -I/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/zlib-1.2.11-eksallf6cymqkp6pkz6ymzjakqt6bqkx/include -L/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-crbbyudcj22o5onhxloe32v5d2nkwa5w/lib  -L/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/zlib-1.2.11-eksallf6cymqkp6pkz6ymzjakqt6bqkx/lib conftest.c -lhdf5 -lz  -lz -lm >&5
configure:28940: $? = 0
configure:28960: result: yes
configure:28989: checking if lack of zlib caused tests for hdf5 to fail...
configure:29314: checking hdf5.h usability
configure:29331: /home/travis/build/junghans/proxy-ci/spack/lib/spack/env/gcc/gcc -c -fPIC -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -Wdeclaration-after-statement -I/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-crbbyudcj22o5onhxloe32v5d2nkwa5w/include  -I/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/zlib-1.2.11-eksallf6cymqkp6pkz6ymzjakqt6bqkx/include conftest.c >&5
In file included from /home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-crbbyudcj22o5onhxloe32v5d2nkwa5w/include/hdf5.h:22:0,
                 from conftest.c:91:
/home/travis/build/junghans/proxy-ci/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-crbbyudcj22o5onhxloe32v5d2nkwa5w/include/H5public.h:59:20: fatal error: mpi.h: No such file or directory
 #   include <mpi.h>
                    ^
compilation terminated.
configure:29338: $? = 1
configure: failed program was:
| /* confdefs.h.  */
| #define PACKAGE_NAME "silo"
| #define PACKAGE_TARNAME "silo"
| #define PACKAGE_VERSION "4.10.2"
| #define PACKAGE_STRING "silo 4.10.2"
| #define PACKAGE_BUGREPORT "visit-users@elist.ornl.gov"
| #define PACKAGE "silo"
| #define VERSION "4.10.2"
| #define FC_FUNC(name,NAME) name ## _
| #define FC_FUNC_(name,NAME) name ## _
| #define STDC_HEADERS 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_SYS_STAT_H 1
.
.
.
```

So, it looks like Silo's configure is failing when looking for HDF5's header file looking for mpi.h. 

Silo does not use or need a parallel HDF5 (e.g. an HDF5 that is built with MPI support). Typically, Silo is built against a serial HDF5 install. But, if you do indeed build Silo against an HDF5 that has MPI support (which I believe should work fine), then you need to tell Silo's configure where to find the MPI header files (via CFLAGS) and MPI library files (via LDFLAGS). 

Note that an MPI-enabled HDF5 installations **does not** adequately support serial application linkgage (e.g. MPI-free linkage). Once you head down the MPI path with HDF5, then all consumers of that install need to know where MPI is regardless of whether those consumers are indeed serial or parallel. This is one of the issues I have been working with HDF5 team to correct.

File: var/spack/repos/builtin/packages/silo/package.py [hunk 1]
```diff
@@ -64,6 +64,11 @@ def install(self, spec, prefix):
                 'CXXFLAGS={0}'.format(self.compiler.pic_flag),
                 'FCFLAGS={0}'.format(self.compiler.pic_flag)]
 
+        # H5public.h includes mpi.h directly
+        if 'hdf5+mpi' in spec:
+            config_args.append(
+                'CPPFLAGS=-I{0}'.format(self.spec['mpi'].prefix.include))
```
Comment (markcmiller86) [inline var/spack/repos/builtin/packages/silo/package.py]: I think this will get you by the configuration failure but when Silo goes to build any of its executables (tools or tests), that will fail due to also needing LDFLAGS. So, I think you also need to add something like so...

```
LDFLAGS=-L{0} -lmpi'.(format(self.spec['mpi'].prefix.lib)
```

File: var/spack/repos/builtin/packages/silo/package.py [hunk 2]
```diff
@@ -64,6 +64,11 @@ def install(self, spec, prefix):
                 'CXXFLAGS={0}'.format(self.compiler.pic_flag),
                 'FCFLAGS={0}'.format(self.compiler.pic_flag)]
 
+        # H5public.h includes mpi.h directly
```
Comment (markcmiller86) [inline var/spack/repos/builtin/packages/silo/package.py]: This comment about HDF5's public header file is true *only* for installs of HDF5 with parallel support.

File: var/spack/repos/builtin/packages/silo/package.py (hunk 1 above)
Comment (markcmiller86) [reply var/spack/repos/builtin/packages/silo/package.py]: I think the upshot from this is that Silo's configure could be a bit smarter about whether its been asked to use serial or parallel install of HDF5.

Comment (junghans) [reply var/spack/repos/builtin/packages/silo/package.py]: It works even without `LDFLAGS`.

Comment (markcmiller86) [reply var/spack/repos/builtin/packages/silo/package.py]: Hmm...are you building Silo tools and/or tests? I can't think their links will complete using an MPI-enabled HDF5 without causing, somehow, -lmpi to appear on the linker command-line. Are you using MPI compiler wrappers by chance?

Comment (junghans) [reply var/spack/repos/builtin/packages/silo/package.py]: I was trying to use `C{C,XX}=mpic{,xx}`, but that led to some linking error.

Anyhow, I think that is really not an silo issue, but an spack CC issue, which assumes that external package follow a standard layout, which isn't the case on Debian/Ubuntu.

**Comment by junghans** [comment]
Actual issue tracked in #7268.

**Comment by junghans** [comment]
I think we still need something like that to fix #7268, but it should be via the header property as introduces in #7267.

File: var/spack/repos/builtin/packages/silo/package.py (hunk 1 above)
Comment (junghans) [reply var/spack/repos/builtin/packages/silo/package.py]: Using `spec["mpi"].headers.directories` (as defined in #7267) should fix it.

**Comment by markcmiller86** [comment]
So, I am not sure where we are at now on this ticket. But, I feel like change Silo package to deal with possibility that someone may use it with a parallel HDF5 is only partially supportable. Silo is a *serial* I/O library. It doesn't use MPI and so should not be used with a parallel HDF5. So, maybe we should just have ``depends_on('hdf5~mpi')`` by fiat. That would also fix this problem and is more truthful in what Silo actually needs, a non-MPI HDF5.

**Comment by junghans** [comment]
I am fine with that, too!

**Comment by markcmiller86** [comment]
Ok, I am fine with you making that change and submitting the solution to this bug that way.

In meantime, I can think about how best to adjust Silo's configure to maybe be more flexible about the fact that it could be used against an MPI-enabled HDF5 installation ;)

**Comment by junghans** [comment]
@markcmiller86  See #7275
